ldurniat/berry.lua

```diff
@@ -447,484 +468,553 @@ end
 -- @param tilesetsDirectory The path to tilesets.
 -- @return The newly created map.
 ------------------------------------------------------------------------------------------------
-function M.new( filename, tilesetsDirectory )
-
-	tilesetsDirectory = tilesetsDirectory and tilesetsDirectory .. '/' or ''
 
+function Map:new( filename, tilesetsDirectory )
 	-- Read map file
     local data = json.decodeFile( system.pathForFile( filename, system.ResourceDirectory ) )
 
-    -- Create group containg all objects including another groups
-    local map  = display.newGroup()
+	local map = setupDisplayGroup( self )
```

Comment:
Why don't  you  add methods to `map` object (like in Ponytiled) instead using 'setupDisplayGroup' function?

Your method :
```
local Map = {}

function Map:new() 
    ...
    local map = setupDisplayGroup( self )
   ...
end
...
function Map:sort()
     ...
end 
```

vs 

Ponytiled method : 

```
local Map = {}

function Map.new() 
    ...
    local map = setupDisplayGroup( self )
   ...
   function map:sort()
     ...
   end

end 
```

Reply:
So I'd say *most* lua libraries in general setup their functions this way: (including ponytiled)

```lua
Library:new()
end

Library:doThis()
end

Library:doThat()
end

Library:doSomething()
end

obj = Library:new(...)
-- obj now has all of Library's methods
```

The way it was initially setup was quite unorthodox.  (I think it's the first time I've ever seen an object inherit its methods that way to be quite honest)

```lua
Library:new()
	obj = ...

	obj:doThis()
	end

	obj:doThat()
	end

	obj:doSomething()
	end
end
```

There was a valid reason for this.  Mainly because you cannot use `setmetatable` on an object that is a display group.  This is due to a display group already having it's own metatable, so the methods need to be inheritted some other way.  One way to get around this is the way it was written initially.  Another way is to just have a `setup` function that attaches the methods to the object like I did.  This is actually what a lot of OO libaries do (middleclass/30log)

As I said in a previous comment, nesting loops too deep is a bad practice.  This problem is magnified if the function itself is large.  Both the deepness and largness were present with the way it was written.  My changes were to resolve this.

On top of all that, I was having issues refactoring the tile and object creation into their own functions with the old way the `M.new` was written.  The order of events for the methods being attacthed caused some unexpected behavior.  

```lua
function M.new(...)
	local map = display.newGroup()

	for i=1, #data.layers do
		...
		if layer.type == 'objectgroup' then
			...
			for object in pairs(objects) do
				map:createObject(...)  -- this is where it would crash

	map:createObject()  -- this function would not load to be used properly

	map:sort()

	map:extend()

	...
```

Because the function `createObject` was created AFTER the objects were getting initialized this resulted in the library crashing.  This is because the `createObject` function does not exist, it has not been loaded by lua yet.  Compare this to:


```lua
function Map:new(...)
	local map = setupDisplayGroup()

	for i=1, #data.layers do
		...
		if layer.type == 'objectgroup' then
			...
			for object in pairs(objects) do
				map:createObject(...)  -- no crashing  :)
	...

Function Map:createObject() -- function loads when library/module is required 
	...
end

Function Map:sort()
	...
end

Function Map:extend()
	...
end
```

The `createObject` function is loaded already and ready to be used.  Hopefully that makes sense.  It took me a while to understand why it was all happening and how to fix it.  Also the scopes of variables were put into their proper function locations where they were needed.

Reply:
I see now how it should work. I have not built before so robust library like Berry. Again thanks for explanation.